src/otx/engine/engine.py

```diff
@@ -0,0 +1,486 @@
+import torch
+import yaml
+from lightning import Trainer, seed_everything
+from lightning.pytorch.cli import instantiate_class
+from rich.console import Console
+from rich.table import Table
+
+from otx.core.config.device import DeviceConfig
+from otx.core.config.engine import EngineConfig
+from otx.core.data.module import OTXDataModule
+from otx.core.model.entity.base import OTXModel
+from otx.core.model.module.base import OTXLitModule
+from otx.core.types.device import OTXDeviceType
+from otx.core.types.task import OTXTaskType
+from otx.core.utils.auto_configuration import AutoConfigurator, load_model_configs
+from otx.core.utils.cache import TrainerArgumentsCache
+from otx.core.utils.instantiators import (
+    instantiate_callbacks,
+    instantiate_loggers,
+    partial_instantiate_class,
+)
+
+if TYPE_CHECKING:
+    from lightning import Callback
+    from lightning.pytorch.cli import LRSchedulerCallable, OptimizerCallable
+    from lightning.pytorch.loggers import Logger
+    from lightning.pytorch.utilities.types import EVAL_DATALOADERS
+    from pytorch_lightning.trainer.connectors.accelerator_connector import _PRECISION_INPUT
+
+
+LITMODULE_PER_TASK = {
+    OTXTaskType.MULTI_CLASS_CLS: "otx.core.model.module.classification.OTXMulticlassClsLitModule",
+    OTXTaskType.MULTI_LABEL_CLS: "otx.core.model.module.classification.OTXMultilabelClsLitModule",
+    OTXTaskType.DETECTION: "otx.core.model.module.detection.OTXDetectionLitModule",
+    OTXTaskType.INSTANCE_SEGMENTATION: "otx.core.model.module.instance_segmentation.OTXInstanceSegLitModule",
+    OTXTaskType.SEMANTIC_SEGMENTATION: "otx.core.model.module.segmentation.OTXSegmentationLitModule",
+    OTXTaskType.ACTION_CLASSIFICATION: "otx.core.model.module.action_classification.OTXActionClsLitModule",
+    OTXTaskType.ACTION_DETECTION: "otx.core.model.module.action_detection.OTXActionDetLitModule",
+}
+
+
+def list_models() -> str:
+    """Returns a list of model names for OTX.
+
+    Returns:
+        str: a table of list model.
+    """
+    table = Table(title="List model of OTX")
+    table.add_column("Task", justify="left", style="yellow")
+    table.add_column("Model Name", justify="left", style="cyan")
+    table.add_column("Config Path", justify="left", style="white")
+    task_list = [task.value for task in LITMODULE_PER_TASK]
+    for task in task_list:
+        model_dict = load_model_configs(task)
+        for model_name, model_path in model_dict.items():
+            table.add_row(task, model_name, model_path)
+    console = Console()
+    with console.capture() as capture:
+        console.print(table, end="")
+    return capture.get()
+
+
+class Engine:
+    """OTX Engine.
+
+    This class defines the common interface for OTX, including methods for training and testing.
+
+    Example:
+    >>> engine = Engine(
+        data_root="dataset/path",
+        task="MULTI_CLASS_CLS",
+    )
+    """
+
+    def __init__(
+        self,
+        *,
+        data_root: str | Path,
```

Comment:
Could this argument be made optional? If the user passes a datamodule object as the `datamodule` argument, we technically don't need `data_root` right?

Reply:
Yes, that's right. I must have missed it, I'll fix it, thanks.